packages/sources/eth-beacon/src/endpoint/balance.ts

```diff
@@ -58,101 +55,82 @@ export const execute: ExecuteWithConfig<Config> = async (request, _, config) =>
-      balances,
-      result: balances,
-    },
-  }
-
-  return Requester.success(jobRunID, result, config.verbose)
+  return await queryInBatches(jobRunID, config, stateId, addresses, validatorStatus)
 }
 
 interface StateResponseSchema {
   execution_optimistic: false
-  data: {
-    index: string
-    balance: string
-    status: string
-    validator: {
-      pubkey: string
-      withdrawal_credentials: string
-      effective_balance: string
-      slashed: boolean
-      activation_eligibility_epoch: string
-      activation_epoch: string
-      exit_epoch: string
-      withdrawable_epoch: string
-    }
+  data: ValidatorState[]
+}
+
+interface ValidatorState {
+  index: string
+  balance: string
+  status: string
+  validator: {
+    pubkey: string
+    withdrawal_credentials: string
+    effective_balance: string
+    slashed: boolean
+    activation_eligibility_epoch: string
+    activation_epoch: string
+    exit_epoch: string
+    withdrawable_epoch: string
   }
 }
 
-const queryWithState = async (
+interface BalanceResponse {
+  address: string
+  balance: string
+}
+
+const queryInBatches = async (
   jobRunID: string,
   config: Config,
   stateId: string,
   addresses: Address[],
   validatorStatus: string[],
 ) => {
-  const url = `/eth/v1/beacon/states/${stateId}/validators/`
+  const url = `/eth/v1/beacon/states/${stateId}/validators`
+  const statusList = validatorStatus?.join(',')
+  const batchSize = Number(config.adapterSpecificParams?.batchSize) || DEFAULT_BATCH_SIZE
+  const batchedAddresses = []
+  // Separate the address set into the specified batch size
+  // Add the batches as comma-separated lists to a new list used to make the requests
+  for (let i = 0; i < addresses.length / batchSize; i++) {
+    batchedAddresses.push(
+      addresses
+        .slice(i * batchSize, i * batchSize + batchSize)
+        .map(({ address }) => address)
+        .join(','),
+    )
+  }
 
+  // Make request to beacon API for every batch
   const responses = await Promise.all(
-    addresses.map(async ({ address }) => {
-      const options = { ...config.api, url: url + address }
-      try {
-        return await Requester.request<StateResponseSchema>(options)
-      } catch {
-        // If address is an invalid validator on the beacon chain, swallow 404 and return a 0 balance
-        return {
-          data: {
-            execution_optimistic: false,
-            data: {
-              index: '0',
-              balance: '0',
-              status: 'failed',
-              validator: {
-                pubkey: address,
-                withdrawal_credentials: '',
-                effective_balance: '0',
-                slashed: false,
-                activation_eligibility_epoch: '',
-                activation_epoch: '',
-                exit_epoch: '',
-                withdrawable_epoch: '',
-              },
-            },
-          } as StateResponseSchema,
-        }
+    batchedAddresses.map((address) => {
```

Comment:
Is there any concern around sending all the batches at the same time?

Reply:
There wasn't at least not at the scale of even the upper limit test. For reliability, should I change it to one at a time?

Reply:
If the responses are quick enough, I think this is fine

Reply:
Ya they're pretty speedy since they're in small batches